File: client/packages/vance-face/src/document/graphCodec.ts
// Codec for `kind: graph` documents — parses an on-disk body into a
// typed GraphDocument and serializes it back. JSON and YAML only;
// markdown is intentionally not supported (see
// `specification/doc-kind-graph.md` §3.3).
//
// Data model: top-level `nodes` and `edges` arrays. Edges are
// first-class entities with their own metadata slot — matches the
// convention used by Cytoscape, GraphML/GEXF, vue-flow internally,
// and the JSON Graph Spec. Edges that point to unknown ids are
// preserved across round-trip but the renderer skips them.

import {
  dumpYamlMultiDoc,
  mergeYamlMultiDoc,
  unwrapJsonMeta,
  wrapJsonMeta,
} from './documentHeaderCodec';

export interface GraphPosition {
  x: number;
  y: number;
}

export interface GraphNode {
  id: string;
  label?: string;
  color?: string;
  position?: GraphPosition;
  /** Unknown per-node fields, preserved across round-trip. */
  extra: Record<string, unknown>;
}

export interface GraphEdge {
  /** Optional stable id. When missing on disk we synthesise
   *  {@code `${source}->${target}`} so the renderer always has one,
   *  but the synthetic value is not written back unless the user
   *  explicitly set it (round-trip preserves the original on-disk
   *  shape). */
  id?: string;
  source: string;
  target: string;
  label?: string;
  color?: string;
  /** Unknown per-edge fields, preserved across round-trip. */
  extra: Record<string, unknown>;
}

export interface GraphConfig {
  /** When true, the renderer draws arrowheads. Default true. */
  directed: boolean;
}

export interface GraphDocument {
  /** Always `'graph'` for graph documents. */
  kind: string;
  graph: GraphConfig;
  nodes: GraphNode[];
  edges: GraphEdge[];
  /** Unknown top-level fields. Re-emitted verbatim on save. */
  extra: Record<string, unknown>;
}

export class GraphCodecError extends Error {
  constructor(message: string, public override readonly cause?: unknown) {
    super(message);
    this.name = 'GraphCodecError';
  }
}

// ── MIME helpers ─────────────────────────────────────────────────────

function isJson(mime: string): boolean {
  return mime === 'application/json';
}
function isYaml(mime: string): boolean {
  return mime === 'application/yaml'
    || mime === 'application/x-yaml'
    || mime === 'text/yaml'
    || mime === 'text/x-yaml';
}

// ── Public API ───────────────────────────────────────────────────────

export function parseGraph(body: string, mimeType: string): GraphDocument {
  if (isJson(mimeType)) return parseGraphJson(body);
  if (isYaml(mimeType)) return parseGraphYaml(body);
  throw new GraphCodecError(`Unsupported mime type for graph: ${mimeType}`);
}

export function serializeGraph(doc: GraphDocument, mimeType: string): string {
  if (isJson(mimeType)) return serializeGraphJson(doc);
  if (isYaml(mimeType)) return serializeGraphYaml(doc);
  throw new GraphCodecError(`Unsupported mime type for graph: ${mimeType}`);
}

/** Whether the codec can handle this mime type — used by the editor
 *  to decide whether to offer the Graph tab at all. Markdown is
 *  intentionally excluded; spec §3.3. */
export function isGraphMime(mimeType: string | null | undefined): boolean {
  if (!mimeType) return false;
  return isJson(mimeType) || isYaml(mimeType);
}

/** Build a fresh empty node with a generated id. The caller passes
 *  the existing node ids so we pick the smallest free `node_<N>`. */
export function emptyNode(existingIds: Iterable<string>): GraphNode {
  const taken = new Set(existingIds);
  let n = 1;
  while (taken.has(`node_${n}`)) n++;
  return {
    id: `node_${n}`,
    extra: {},
  };
}

/** Synthesise the renderer-friendly id for an edge. The on-disk
 *  representation may omit {@code id}; the editor / renderer needs a
 *  unique handle to track the edge across renders. Falls back to
 *  {@code source->target} which is also what the legacy migration
 *  produces. */
export function edgeKey(edge: GraphEdge): string {
  return edge.id && edge.id.length > 0
    ? edge.id
    : `${edge.source}->${edge.target}`;
}

// ── JSON ─────────────────────────────────────────────────────────────

function parseGraphJson(body: string): GraphDocument {
  if (body.trim() === '') return emptyDoc();
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch (e) {
    throw new GraphCodecError('Invalid JSON: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  if (!isObject(parsed)) {
    throw new GraphCodecError('Top-level JSON must be an object');
  }
  return promoteToGraphDocument(unwrapJsonMeta(parsed));
}

function serializeGraphJson(doc: GraphDocument): string {
  const body = buildOnDiskBody(doc);
  return JSON.stringify(wrapJsonMeta(doc.kind || 'graph', body), null, 2) + '\n';
}

// ── YAML ─────────────────────────────────────────────────────────────

function parseGraphYaml(body: string): GraphDocument {
  if (body.trim() === '') return emptyDoc();
  let merged: Record<string, unknown>;
  try {
    merged = mergeYamlMultiDoc(body);
  } catch (e) {
    throw new GraphCodecError('Invalid YAML: ' + (e instanceof Error ? e.message : String(e)), e);
  }
  return promoteToGraphDocument(merged);
}

function serializeGraphYaml(doc: GraphDocument): string {
  return dumpYamlMultiDoc(doc.kind || 'graph', buildOnDiskBody(doc));
}

// ── Shared promotion + writeback ────────────────────────────────────

function emptyDoc(): GraphDocument {
  return {
    kind: 'graph',
    graph: { directed: true },
    nodes: [],
    edges: [],
    extra: {},
  };
}

function promoteToGraphDocument(obj: Record<string, unknown>): GraphDocument {
  const kind = typeof obj.kind === 'string' ? obj.kind : '';
  const graph = promoteConfig(obj.graph);

  // Promote nodes first so legacy `node.edges` migration can see the
  // valid node-id set when synthesising top-level edges.
  const { nodes, legacyEdges } = promoteNodes(obj.nodes);
  const explicitEdges = promoteEdges(obj.edges);
  // Combined: explicit top-level edges win over legacy node-edges of
  // the same shape. Legacy edges only fire for old-format documents
  // that hadn't been resaved since the spec change; new saves never
  // emit `edges` on a node, so this is a one-time migration.
  const edges = mergeEdges(explicitEdges, legacyEdges);

  const { kind: _k, graph: _g, nodes: _n, edges: _e, ...extra } = obj;
  return { kind, graph, nodes, edges, extra };
}

function promoteConfig(raw: unknown): GraphConfig {
  if (isObject(raw) && typeof raw.directed === 'boolean') {
    return { directed: raw.directed };
  }
  return { directed: true };
}

interface NodesPromoteResult {
  nodes: GraphNode[];
  legacyEdges: GraphEdge[];
}

function promoteNodes(raw: unknown): NodesPromoteResult {
  if (!Array.isArray(raw)) return { nodes: [], legacyEdges: [] };
  const nodes: GraphNode[] = [];
  const legacyEdges: GraphEdge[] = [];
  const seenIds = new Set<string>();
  for (const r of raw) {
    if (!isObject(r)) continue;
    const idRaw = r.id;
    if (typeof idRaw !== 'string') continue;
    const id = idRaw.trim();
    if (!id) continue;
    if (seenIds.has(id)) {
      throw new GraphCodecError(`Duplicate node id: ${id}`);
    }
    seenIds.add(id);
    const node: GraphNode = { id, extra: {} };
    if (typeof r.label === 'string') node.label = r.label;
    if (typeof r.color === 'string' && r.color) node.color = r.color;
    const pos = promotePosition(r.position);
    if (pos) node.position = pos;

    // Backward-compat: legacy out-edge string-list. Lift each entry
    // into a top-level edge; the new format never writes node.edges,
    // so this only triggers for old documents.
    if (Array.isArray(r.edges)) {
      for (const target of r.edges) {
        if (typeof target === 'string' && target.trim()) {
          legacyEdges.push({
            source: id,
            target: target.trim(),
            extra: {},
          });
        }
      }
    }

    for (const [k, v] of Object.entries(r)) {
      if (k === 'id' || k === 'label' || k === 'color' || k === 'position' || k === 'edges') continue;
      node.extra[k] = v;
    }
    nodes.push(node);
  }
  return { nodes, legacyEdges };
}

function promoteEdges(raw: unknown): GraphEdge[] {
  if (!Array.isArray(raw)) return [];
  const out: GraphEdge[] = [];
  for (const r of raw) {
    if (!isObject(r)) continue;
    const sourceRaw = r.source;
    const targetRaw = r.target;
    if (typeof sourceRaw !== 'string' || typeof targetRaw !== 'string') continue;
    const source = sourceRaw.trim();
    const target = targetRaw.trim();
    if (!source || !target) continue;
    const edge: GraphEdge = { source, target, extra: {} };
    if (typeof r.id === 'string' && r.id.trim()) edge.id = r.id.trim();
    if (typeof r.label === 'string') edge.label = r.label;
    if (typeof r.color === 'string' && r.color) edge.color = r.color;
    for (const [k, v] of Object.entries(r)) {
      if (k === 'id' || k === 'source' || k === 'target' || k === 'label' || k === 'color') continue;
      edge.extra[k] = v;
    }
    out.push(edge);
  }
  return out;
}

/** Combine explicit top-level edges with legacy node-edges. Explicit
 *  wins on collision (same id, or same source-target without ids). */
function mergeEdges(explicit: GraphEdge[], legacy: GraphEdge[]): GraphEdge[] {
  if (legacy.length === 0) return explicit;
  const seen = new Set<string>();
  for (const e of explicit) seen.add(edgeKey(e));
  const out = [...explicit];
  for (const e of legacy) {
    const key = edgeKey(e);
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(e);
  }
  return out;
}

function promotePosition(raw: unknown): GraphPosition | undefined {
  if (!isObject(raw)) return undefined;
  const x = typeof raw.x === 'number' && Number.isFinite(raw.x) ? raw.x : undefined;
  const y = typeof raw.y === 'number' && Number.isFinite(raw.y) ? raw.y : undefined;
  if (x === undefined || y === undefined) return undefined;
  return { x, y };
}

/** Top-level body keys (everything except `kind`, which lives in
 *  the `$meta` wrapper). */
function buildOnDiskBody(doc: GraphDocument): Record<string, unknown> {
  return {
    graph: { directed: doc.graph?.directed ?? true },
    nodes: doc.nodes.map(nodeToObject),
    edges: doc.edges.map(edgeToObject),
    ...doc.extra,
  };
}

function nodeToObject(node: GraphNode): Record<string, unknown> {
  const obj: Record<string, unknown> = { id: node.id };
  if (node.label !== undefined) obj.label = node.label;
  if (node.color !== undefined) obj.color = node.color;
  if (node.position !== undefined) {
    obj.position = { x: node.position.x, y: node.position.y };
  }
  for (const [k, v] of Object.entries(node.extra)) {
    if (!(k in obj)) obj[k] = v;
  }
  return obj;
}

function edgeToObject(edge: GraphEdge): Record<string, unknown> {
  const obj: Record<string, unknown> = {};
  // `id` is optional on disk: only written when the user has set
  // one explicitly. Synthesised ids (from edgeKey() fallback) live
  // only in memory and don't get persisted, keeping the on-disk
  // shape minimal.
  if (edge.id !== undefined) obj.id = edge.id;
  obj.source = edge.source;
  obj.target = edge.target;
  if (edge.label !== undefined) obj.label = edge.label;
  if (edge.color !== undefined) obj.color = edge.color;
  for (const [k, v] of Object.entries(edge.extra)) {
    if (!(k in obj)) obj[k] = v;
  }
  return obj;
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

File: server/vance-shared/src/main/java/de/mhus/vance/shared/document/kind/GraphCodec.java
package de.mhus.vance.shared.document.kind;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.jspecify.annotations.Nullable;
import tools.jackson.core.JacksonException;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

/**
 * Parser and serialiser for {@code kind: graph} document bodies —
 * mirrors {@code graphCodec.ts}. JSON and YAML only; markdown is
 * intentionally not supported.
 *
 * <p>Reads both the canonical top-level {@code edges} array and the
 * legacy {@code node.edges: string[]} form (which gets lifted on
 * read; {@code mergeEdges} prefers explicit top-level entries on
 * collision). Writes only the canonical form.
 */
public final class GraphCodec {

    private static final ObjectMapper JSON = new ObjectMapper();
    private static final TypeReference<LinkedHashMap<String, Object>> JSON_MAP =
            new TypeReference<>() {};

    private GraphCodec() {
        // utility class
    }

    public static GraphDocument parse(String body, @Nullable String mimeType) {
        if (isJson(mimeType)) return parseJson(body);
        if (isYaml(mimeType)) return parseYaml(body);
        throw new KindCodecException("Unsupported mime type for graph: " + mimeType);
    }

    public static String serialize(GraphDocument doc, @Nullable String mimeType) {
        if (isJson(mimeType)) return serializeJson(doc);
        if (isYaml(mimeType)) return serializeYaml(doc);
        throw new KindCodecException("Unsupported mime type for graph: " + mimeType);
    }

    public static boolean supports(@Nullable String mimeType) {
        return isJson(mimeType) || isYaml(mimeType);
    }

    // ── Mime ───────────────────────────────────────────────────────

    private static boolean isJson(@Nullable String mime) {
        return "application/json".equals(mime);
    }
    private static boolean isYaml(@Nullable String mime) {
        return "application/yaml".equals(mime)
                || "application/x-yaml".equals(mime)
                || "text/yaml".equals(mime)
                || "text/x-yaml".equals(mime);
    }

    // ── JSON / YAML ────────────────────────────────────────────────

    private static GraphDocument parseJson(String body) {
        if (body.isBlank()) return GraphDocument.empty();
        Map<String, Object> parsed;
        try {
            parsed = JSON.readValue(body, JSON_MAP);
        } catch (JacksonException e) {
            throw new KindCodecException("Invalid JSON: " + e.getOriginalMessage(), e);
        }
        if (parsed == null) throw new KindCodecException("Top-level JSON must be an object");
        return promoteToDocument(KindHeaderCodec.unwrapJsonMeta(parsed));
    }

    private static GraphDocument parseYaml(String body) {
        if (body.isBlank()) return GraphDocument.empty();
        return promoteToDocument(KindHeaderCodec.mergeYamlMultiDoc(body));
    }

    private static String serializeJson(GraphDocument doc) {
        Map<String, Object> wrapped = KindHeaderCodec.wrapJsonMeta(canonicalKind(doc), buildBody(doc));
        try {
            return JSON.writerWithDefaultPrettyPrinter().writeValueAsString(wrapped) + "\n";
        } catch (JacksonException e) {
            throw new KindCodecException("Failed to write JSON: " + e.getOriginalMessage(), e);
        }
    }

    private static String serializeYaml(GraphDocument doc) {
        return KindHeaderCodec.dumpYamlMultiDoc(canonicalKind(doc), buildBody(doc));
    }

    // ── Promotion ──────────────────────────────────────────────────

    private static GraphDocument promoteToDocument(Map<String, Object> obj) {
        Object kindRaw = obj.get("kind");
        String kind = (kindRaw instanceof String s) ? s : "";
        GraphConfig graph = promoteConfig(obj.get("graph"));

        // Promote nodes first so legacy node.edges migration can
        // synthesise top-level edges referencing valid node ids.
        NodesPromote nodesP = promoteNodes(obj.get("nodes"));
        List<GraphEdge> explicitEdges = promoteEdges(obj.get("edges"));
        List<GraphEdge> edges = mergeEdges(explicitEdges, nodesP.legacyEdges);

        Map<String, Object> extra = new LinkedHashMap<>(obj);
        extra.remove("kind");
        extra.remove("graph");
        extra.remove("nodes");
        extra.remove("edges");
        return new GraphDocument(kind.isEmpty() ? "graph" : kind, graph, nodesP.nodes, edges, extra);
    }

    private static GraphConfig promoteConfig(@Nullable Object raw) {
        if (raw instanceof Map<?, ?> map && map.get("directed") instanceof Boolean b) {
            return new GraphConfig(b);
        }
        return GraphConfig.defaults();
    }

    private record NodesPromote(List<GraphNode> nodes, List<GraphEdge> legacyEdges) {}

    private static NodesPromote promoteNodes(@Nullable Object raw) {
        List<GraphNode> nodes = new ArrayList<>();
        List<GraphEdge> legacyEdges = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return new NodesPromote(nodes, legacyEdges);
        Set<String> seen = new LinkedHashSet<>();
        for (Object r : list) {
            if (!(r instanceof Map<?, ?> map)) continue;
            Object idRaw = map.get("id");
            if (!(idRaw instanceof String idS)) continue;
            String id = idS.trim();
            if (id.isEmpty()) continue;
            if (seen.contains(id)) {
                throw new KindCodecException("Duplicate node id: " + id);
            }
            seen.add(id);

            String label = (map.get("label") instanceof String ls) ? ls : null;
            String color = (map.get("color") instanceof String cs && !cs.isEmpty()) ? cs : null;
            GraphPosition position = promotePosition(map.get("position"));

            // Legacy node.edges: lift to top-level edges.
            if (map.get("edges") instanceof List<?> le) {
                for (Object e : le) {
                    if (e instanceof String es && !es.trim().isEmpty()) {
                        legacyEdges.add(new GraphEdge(null, id, es.trim(),
                                null, null, new LinkedHashMap<>()));
                    }
                }
            }

            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if ("id".equals(key) || "label".equals(key) || "color".equals(key)
                        || "position".equals(key) || "edges".equals(key)) continue;
                extra.put(key, e.getValue());
            }
            nodes.add(new GraphNode(id, label, color, position, extra));
        }
        return new NodesPromote(nodes, legacyEdges);
    }

    private static List<GraphEdge> promoteEdges(@Nullable Object raw) {
        List<GraphEdge> out = new ArrayList<>();
        if (!(raw instanceof List<?> list)) return out;
        for (Object r : list) {
            if (!(r instanceof Map<?, ?> map)) continue;
            Object srcRaw = map.get("source");
            Object tgtRaw = map.get("target");
            if (!(srcRaw instanceof String ss) || !(tgtRaw instanceof String ts)) continue;
            String source = ss.trim();
            String target = ts.trim();
            if (source.isEmpty() || target.isEmpty()) continue;
            String id = (map.get("id") instanceof String is && !is.trim().isEmpty()) ? is.trim() : null;
            String label = (map.get("label") instanceof String ls) ? ls : null;
            String color = (map.get("color") instanceof String cs && !cs.isEmpty()) ? cs : null;
            Map<String, Object> extra = new LinkedHashMap<>();
            for (Map.Entry<?, ?> e : map.entrySet()) {
                if (!(e.getKey() instanceof String key)) continue;
                if ("id".equals(key) || "source".equals(key) || "target".equals(key)
                        || "label".equals(key) || "color".equals(key)) continue;
                extra.put(key, e.getValue());
            }
            out.add(new GraphEdge(id, source, target, label, color, extra));
        }
        return out;
    }

    private static List<GraphEdge> mergeEdges(List<GraphEdge> explicit, List<GraphEdge> legacy) {
        if (legacy.isEmpty()) return explicit;
        Set<String> seen = new LinkedHashSet<>();
        for (GraphEdge e : explicit) seen.add(e.key());
        List<GraphEdge> out = new ArrayList<>(explicit);
        for (GraphEdge e : legacy) {
            String key = e.key();
            if (seen.contains(key)) continue;
            seen.add(key);
            out.add(e);
        }
        return out;
    }

    private static @Nullable GraphPosition promotePosition(@Nullable Object raw) {
        if (!(raw instanceof Map<?, ?> map)) return null;
        Object xR = map.get("x");
        Object yR = map.get("y");
        if (!(xR instanceof Number xn) || !(yR instanceof Number yn)) return null;
        double x = xn.doubleValue();
        double y = yn.doubleValue();
        if (!Double.isFinite(x) || !Double.isFinite(y)) return null;
        return new GraphPosition(x, y);
    }

    // ── Body builder ───────────────────────────────────────────────

    private static Map<String, Object> buildBody(GraphDocument doc) {
        Map<String, Object> body = new LinkedHashMap<>();
        Map<String, Object> graph = new LinkedHashMap<>();
        graph.put("directed", doc.graph().directed());
        body.put("graph", graph);
        body.put("nodes", nodesToList(doc.nodes()));
        body.put("edges", edgesToList(doc.edges()));
        for (Map.Entry<String, Object> e : doc.extra().entrySet()) {
            if (!body.containsKey(e.getKey())) body.put(e.getKey(), e.getValue());
        }
        return body;
    }

    private static List<Map<String, Object>> nodesToList(List<GraphNode> nodes) {
        List<Map<String, Object>> out = new ArrayList<>(nodes.size());
        for (GraphNode n : nodes) {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", n.id());
            if (n.label() != null) m.put("label", n.label());
            if (n.color() != null) m.put("color", n.color());
            if (n.position() != null) {
                Map<String, Object> pos = new LinkedHashMap<>();
                pos.put("x", n.position().x());
                pos.put("y", n.position().y());
                m.put("position", pos);
            }
            for (Map.Entry<String, Object> e : n.extra().entrySet()) {
                if (!m.containsKey(e.getKey())) m.put(e.getKey(), e.getValue());
            }
            out.add(m);
        }
        return out;
    }

    private static List<Map<String, Object>> edgesToList(List<GraphEdge> edges) {
        List<Map<String, Object>> out = new ArrayList<>(edges.size());
        for (GraphEdge e : edges) {
            Map<String, Object> m = new LinkedHashMap<>();
            if (e.id() != null) m.put("id", e.id());
            m.put("source", e.source());
            m.put("target", e.target());
            if (e.label() != null) m.put("label", e.label());
            if (e.color() != null) m.put("color", e.color());
            for (Map.Entry<String, Object> ex : e.extra().entrySet()) {
                if (!m.containsKey(ex.getKey())) m.put(ex.getKey(), ex.getValue());
            }
            out.add(m);
        }
        return out;
    }

    private static String canonicalKind(GraphDocument doc) {
        return (doc.kind() == null || doc.kind().isBlank()) ? "graph" : doc.kind();
    }
}
